Firestore/core/test/firebase/firestore/local/mutation_queue_test.cc

```diff
@@ -0,0 +1,432 @@
+    }
+
+    DocumentKeySet keys{
+        Key("foo/bar"),
+        Key("foo/baz"),
+    };
+
+    std::vector<MutationBatch> expected{batches[1], batches[2], batches[4]};
+    std::vector<MutationBatch> matches =
+        mutation_queue_->AllMutationBatchesAffectingDocumentKeys(keys);
+
+    ASSERT_EQ(matches, expected);
+  });
+}
+
+TEST_P(MutationQueueTest,
+       AllMutationBatchesAffectingDocumentKeys_handlesOverlap) {
+  persistence_->Run(
+      "test_all_mutation_batches_affecting_document_keys_handlesOverlap", [&] {
+        std::vector<Mutation> group1 = {
+            testutil::SetMutation("foo/bar", Map("a", 1)),
+            testutil::SetMutation("foo/baz", Map("a", 1)),
+        };
+        MutationBatch batch1 = mutation_queue_->AddMutationBatch(
+            Timestamp::Now(), {}, std::move(group1));
+
+        std::vector<Mutation> group2 = {
+            testutil::SetMutation("food/bar", Map("a", 1))};
+        mutation_queue_->AddMutationBatch(Timestamp::Now(), {},
+                                          std::move(group2));
+
+        std::vector<Mutation> group3 = {
+            testutil::SetMutation("foo/bar", Map("b", 1)),
+        };
+        MutationBatch batch3 = mutation_queue_->AddMutationBatch(
+            Timestamp::Now(), {}, std::move(group3));
+
+        DocumentKeySet keys{
+            Key("foo/bar"),
+            Key("foo/baz"),
+        };
+
+        std::vector<MutationBatch> expected{batch1, batch3};
+        std::vector<MutationBatch> matches =
+            mutation_queue_->AllMutationBatchesAffectingDocumentKeys(keys);
+
+        ASSERT_EQ(matches, expected);
+      });
+}
+
+TEST_P(MutationQueueTest, AllMutationBatchesAffectingQuery) {
+  persistence_->Run("test_all_mutation_batches_affecting_query", [&] {
+    std::vector<Mutation> mutations = {
+        testutil::SetMutation("fob/bar", Map("a", 1)),
+        testutil::SetMutation("foo/bar", Map("a", 1)),
+        testutil::PatchMutation("foo/bar", Map("b", 1), {}),
+        testutil::SetMutation("foo/bar/suffix/key", Map("a", 1)),
+        testutil::SetMutation("foo/baz", Map("a", 1)),
+        testutil::SetMutation("food/bar", Map("a", 1)),
+    };
+
+    // Store all the mutations.
+    std::vector<MutationBatch> batches;
+    for (const Mutation& mutation : mutations) {
+      MutationBatch batch =
+          mutation_queue_->AddMutationBatch(Timestamp::Now(), {}, {mutation});
+      batches.push_back(batch);
+    }
+
+    std::vector<MutationBatch> expected = {batches[1], batches[2], batches[4]};
+    core::Query query = Query("foo");
+    std::vector<MutationBatch> matches =
+        mutation_queue_->AllMutationBatchesAffectingQuery(query);
+
+    ASSERT_EQ(matches, expected);
+  });
+}
+
+TEST_P(MutationQueueTest, RemoveMutationBatches) {
+  persistence_->Run("test_remove_mutation_batches", [&] {
+    std::vector<MutationBatch> batches = CreateBatches(10);
+
+    mutation_queue_->RemoveMutationBatch(batches[0]);
```

Comment:
Perhaps use `front`? Applies everywhere in this function where only the first element is removed "in one go".

Reply:
Since this test concerns itself with things in the first few positions by index I've opted not to treat the first specially.